### DISCUSSION TITLE
FreeRTOS命令注册无法执行问题

### DISCUSSION THREAD

**Discussion by bevis-wong** [NevermindZZT/letter-shell#85] [Q&A] [answered]
letter shell我在FreeRTOS运行使用以下注册方式注册获取系统状态，总是无法正常运行。
执行函数为  system_view_state ->  vTaskList( char * pcWriteBuffer ) -> 内部需要动态申请内存pvPortMalloc，却一直无法申请成功。
当我把函数 system_view_state 放到外面正常流程去执行却是可以的。请问是什么原因呢？

``SHELL_EXPORT_CMD(SHELL_CMD_PERMISSION(0)|SHELL_CMD_TYPE(SHELL_TYPE_CMD_MAIN), ps, system_view_state, Show Task State);
//SHELL_EXPORT_CMD(SHELL_CMD_PERMISSION(0)|SHELL_CMD_TYPE(SHELL_TYPE_CMD_FUNC), ps, system_view_state, Show Task State);
//SHELL_EXPORT_CMD(SHELL_CMD_PERMISSION(0)|SHELL_CMD_TYPE(SHELL_TYPE_CMD_FUNC), func, func, Show Task State);

**Comment by NevermindZZT** [comment, marked as answer]
命令注册是没问题的，不知道你这个函数的实现是什么样的，给你一个我实现的命令，可以参考下

```c
/**
 * @brief 打印任务列表
 * 
 */
void showTaskList(void)
{
#if configUSE_TRACE_FACILITY == 1 && configUSE_STATS_FORMATTING_FUNCTIONS == 1
    char *buf = pvPortMalloc(2048);
    if (buf != NULL)
    {
        logPrintln("Name            State   Pri     Stack   Num");
        logPrintln("*******************************************");
        vTaskList(buf);
        logPrintln("%s", buf);
        vPortFree(buf);
    }
#else
    logPrintln("fuction vTaskList is disabled");
#endif
}
SHELL_EXPORT_CMD(
SHELL_CMD_PERMISSION(0)|SHELL_CMD_TYPE(SHELL_TYPE_CMD_FUNC)|SHELL_CMD_DISABLE_RETURN,
ps, showTaskList, show task list);
```

**Reply by bevis-wong** [reply]
非常感谢，我重新下载TAG版本3.0.6测试后，发现可以了，可能跟我使用版本有关，也可能跟我工程有关
